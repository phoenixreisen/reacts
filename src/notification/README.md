# Phoenix Status Notification

[React](https://reactjs.org/)-Komponente für die Darstellung von Status Notification.

Die Komponente ist Teil des [Phoenix Reisen Design-Systems](https://design-system.phoenixreisen.net).

## Demo

https://phoenixreisen.github.io/reacts/notification/

## Installation

[React](https://reactjs.org/) wird benötigt.

```bash
npm install --save @phoenixreisen/reacts
```

### Mehrere Notifications rendern

Nach einem Benutzerereignis (z.B. Speichern oder Löschen) wird ein Notification-Objekt zu einem **gestreamten Array** auszugebender Notifications hinzugefügt. Anschließend wird es der `Notifications`-Komponente als Parameter übergeben.

`Notifications` iteriert über die Liste und rendert entsprechend oft die `Notification`-Komponente mit den jeweiligen Objektdaten. Danach, nach ca. 5 Sekunden, ruft `Notification` die als Parameter übergebene `toggle()`-Funktion auf, die dafür sorgt, dass das jeweilige Notification-Objekt aus der Liste gelöscht wird.

```tsx
import Notifications, { notify } from '@phoenixreisen/reacts/src/notifications';

const ExampleView = () => {

    const submit = () => {
        Promise.resolve('saved!').then(() => {
            notify({
                status: 'success',
                text: 'Erfolgreich gespeichert!',
            });
        });
    };

    return (
        <div>
            <div>Irgendwelcher Content</div>
            <Notifications />
        </div>
    }
};
```

## Test

```bash
npm install
npm test
```

## Demo für Github Page

Mit folgendem Befehl werden alle Demos in den Ordner [../../docs](../../docs) gebaut. Dieser Ordner wird von Github zur Bereitstellung der Demos bzw. Github Pages genutzt.

Möchte man zum Entwickeln nur bestimmte Demos bauen, kann man in der [rollup.config.js](../../rollup.config.js) nicht benötigte Demos auskommentieren, sollte sie vor dem `push` aber wieder einkommentieren.

```bash
npm run compile:demos
```

## Test

```bash
[npm install]
npm run test
```

## Deployment

Deployed & published wird immer die ganze Sammlung. [Siehe hier](../../README.md).