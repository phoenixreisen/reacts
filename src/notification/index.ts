import Notifications from "./notifications.r";

export { NoteObject, NoteProps } from './notification.api';
export { ICONS, SHOWTIMER, STATUS } from './notification.api';
export { notify, unnote, notes, onNoteListChange } from './notification.api';

export { Notifications } from './notifications.r';
export { Notification } from './notification.r';

export default Notifications;