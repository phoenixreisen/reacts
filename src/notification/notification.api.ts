//--- Types -----

export interface NoteObject {
    status?: STATUS,
    text: string
}

export interface NoteProps {
    status?: STATUS,
    toggle: () => void,
    text: string
}


//--- Enums -----

// Namen der FontAwesome Icons für versch. Stati
export const enum ICONS {
    check = 'check',
    info = 'info-circle',
    triangle = 'exclamation-triangle',
}

// Mögliche Stati
export const enum STATUS {
    error = 'error',
    normal = 'normal',
    success = 'success',
}


//--- Variablen & Konstanten -----

// Zeit, bis die Meldung wieder ausgeblendet wird
// mit 100ms extra zur CSS Transition
export const SHOWTIMER = 8100;

// Die Liste aller auszugebenden Notification-Objekte
export const notes = new Set<NoteObject>();

// Liste von callback Funktionen, die bei einer neuen
// oder gelöschten Notification nacheinander aufgerufen werden.
export const onNoteListChange = new Set<() => void>();

// Neue Notification hinzufügen oder löschen.
export const notify = (note?: NoteObject, type: 'add'|'del'= 'add') => {
    if(note) {
        type === 'add' && notes.add(note);
        type === 'del' && notes.delete(note);
    }
    for(const callback of Array.from(onNoteListChange)) {
        callback();
    }
}

// Wrapper-Funktion für obigen Aufruf zum löschen einer Notification.
export const unnote = (note: NoteObject) => {
    notify(note, 'del');
}