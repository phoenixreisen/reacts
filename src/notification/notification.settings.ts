import stream from 'mithril/stream';

//--- Types -----

export interface NoteObject {
    status?: STATUS,
    text: string | Element | JSX.Element | React.Component
}

export interface NoteProps {
    status?: STATUS,
    toggle: () => void,
    text: string | Element | JSX.Element | React.Component
}


//--- Variablen & Konstanten -----

// Zeit, bis die Meldung wieder ausgeblendet wird
// mit 100ms extra zur CSS Transition
export const SHOWTIMER = 8100;

// Die Liste aller auszugebenden Notification-Objekte
export const notes = stream<Array<NoteObject>>([]);


//--- Enums -----

// Namen der FontAwesome Icons für versch. Stati
export const enum ICONS {
    check = 'check',
    info = 'info-circle',
    triangle = 'exclamation-triangle',
}

// Mögliche Stati
export const enum STATUS {
    error = 'error',
    success = 'success',
}