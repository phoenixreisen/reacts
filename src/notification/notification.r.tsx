import { NoteProps, STATUS, ICONS, SHOWTIMER } from './notification.settings';
import React, { useEffect, useState } from 'react';

//--- Ausgabe einer Notification -----

export const Notification = (props: NoteProps) => {
    const [ icon, setIcon ] = useState<string>(null);

    useEffect(() => {
        if(!props.toggle || !props.text) {
            throw 'need at least toggle() and text!';
        }
        setIcon((props.status === STATUS.error)
            ? ICONS.triangle
            : (props.status === STATUS.success)
                ? ICONS.check
                : ICONS.info,
        );
    }, []);

    useEffect(() => {
        const to = setTimeout(() => props.toggle(), SHOWTIMER);
        return () => clearTimeout(to);
    }, []);

    return (
        <article className={`notification notification--${props.status || 'primary'}`}>
            <i className={`fas fa-${icon}`}></i>
            <span>{ props.text }</span>
            <a href="#" className="fas fa-times" onClick={(e) => {
                e?.preventDefault();
                props.toggle();
            }}></a>
        </article>
    );
};

export default Notification;