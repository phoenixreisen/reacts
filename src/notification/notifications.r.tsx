import { NoteObject, notes } from './notification.settings';
import React, { useEffect, useState } from 'react';
import { Notification } from './notification.r';

//--- Ausgabe aller Notifications -----

export const Notifications = () => {
    const [rerender, setRerender] = useState<number>(0);

    useEffect(() => {
        notes.map(() => setRerender(notes().length));
    }, [rerender]);

    return (
        <React.Fragment>
            {notes().map?.((note:NoteObject, index:number) =>
                <Notification
                    key={index}
                    text={note.text}
                    status={note.status}
                    toggle={() => notes(notes().filter(current => current !== note))}
                />,
            )}
        </React.Fragment>
    );
};

export default Notifications;