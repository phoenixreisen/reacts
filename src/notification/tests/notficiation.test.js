/**
 * @jest-environment jsdom
 */
import {render, fireEvent, waitFor, screen, getByText} from '@testing-library/react';
import { STATUS, notes } from '../notification.settings';
import Notifications from '../notifications.r';
import Notification from '../notification.r';
import renderer from 'react-test-renderer';
import React from 'react';

describe('Notification should', () => {

    test('render success correctly', () => {
        const state = { show: true };
        const success = renderer.create(
            <Notification
                status={STATUS.success}
                text={'Erfolgreich gespeichert'}
                toggle={() => state.show = false}
            />
        );
        const snap = success.toJSON();
        expect(snap).toMatchSnapshot();
    });

    test('render info correctly', () => {
        const state = { show: true };
        const success = renderer.create(
            <Notification
                status={STATUS.normal}
                text={'Ich bin ein Bibabutzemann!'}
                toggle={() => state.show = false}
            />
        );
        const snap = success.toJSON();
        expect(snap).toMatchSnapshot();
    });

    test('render error correctly', () => {
        const state = { show: true };
        const success = renderer.create(
            <Notification
                status={STATUS.error}
                text={'WTF?! Ein Error!'}
                toggle={() => state.show = false}
            />
        );
        const snap = success.toJSON();
        expect(snap).toMatchSnapshot();
    });

    test('fail without "text" && "toggle()" parameters', () => {
        const out = console.error;
        console.error = () => {};
        let error = null;
        try {
            renderer.create(<Notification />);
        } catch(e) {
            error = e;
        }
        expect(error).not.toBeNull();
        console.error = out;
    });

    test('render a list of notifications', () => {

        notes().push({ text: "Note 3", status: STATUS.error });
        notes().push({ text: "Note 1", status: STATUS.normal });
        notes().push({ text: "Note 2", status: STATUS.success });

        const comp = render(<Notifications />);

        expect(screen.getByText('Note 1')).not.toBeNull();
        expect(screen.getByText('Note 2')).not.toBeNull();
        expect(screen.getByText('Note 3')).not.toBeNull();
        expect(document.querySelector('.notification--error')).not.toBeNull();
        expect(document.querySelector('.notification--success')).not.toBeNull();
    });
});