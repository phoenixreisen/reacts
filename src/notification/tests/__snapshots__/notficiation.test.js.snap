// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification should render error correctly 1`] = `
<article
  className="notification notification--error"
>
  <i
    className="fas fa-null"
  />
  <span>
    WTF?! Ein Error!
  </span>
  <a
    className="fas fa-times"
    href="#"
    onClick={[Function]}
  />
</article>
`;

exports[`Notification should render info correctly 1`] = `
<article
  className="notification notification--normal"
>
  <i
    className="fas fa-info-circle"
  />
  <span>
    Ich bin ein Bibabutzemann!
  </span>
  <a
    className="fas fa-times"
    href="#"
    onClick={[Function]}
  />
</article>
`;

exports[`Notification should render success correctly 1`] = `
<article
  className="notification notification--success"
>
  <i
    className="fas fa-null"
  />
  <span>
    Erfolgreich gespeichert
  </span>
  <a
    className="fas fa-times"
    href="#"
    onClick={[Function]}
  />
</article>
`;
